crates/pipeline-manager/src/runner/logs_buffer.rs

```diff
@@ -0,0 +1,221 @@
+use std::collections::VecDeque;
+
+/// The LogsBuffer maintains internally a circular buffer of Strings whose
+/// size in byte and number of elements does not exceed the limits.
+/// When appending new log lines (Strings) to the buffer, the limits are
+/// enforced by discarding existing lines if the limits would be exceeded.
+pub struct LogsBuffer {
+    /// Buffer size limit in byte.
+    size_limit_byte: usize,
+    /// Buffer size limit in number of lines.
+    size_limit_num_lines: usize,
+    /// The lines buffer.
+    buffer: VecDeque<String>,
+    /// Current lines buffer size.
+    size_byte: usize,
+    /// Number of lines that have been discarded to enforce size limit.
+    num_discarded_lines: usize,
+}
+
+impl LogsBuffer {
+    /// Construct a new logs buffer.
+    pub fn new(size_limit_byte: usize, size_limit_num_lines: usize) -> Self {
+        Self {
+            size_limit_byte,
+            size_limit_num_lines,
+            buffer: VecDeque::new(),
+            size_byte: 0,
+            num_discarded_lines: 0,
+        }
+    }
+
+    /// Append a new line to the buffer.
+    /// - If the new line exceeds the buffer size limit by itself, all the lines in the buffer and
+    ///   the new line are discarded, leaving an empty buffer.
+    /// - Otherwise, lines are removed from the buffer until the new line will fit. Once there
+    ///   is sufficient space, the line is added to the buffer.
+    pub fn append(&mut self, line: String) {
+        if line.len() > self.size_limit_byte {
+            self.num_discarded_lines += self.buffer.len() + 1;
```

Comment:
Would it be better to truncate such a line to a reasonable length.  In fact, this might be a good strategy for all lines, even if they fit in the log.

Reply:
I see it more as a terminal experience, in which truncating lines also does not happen. The line length would be quite exceptional if it is more than a megabyte by itself -- the log lines are the output of the pipeline executable, as such such long log lines would not occur unless the adapters crate or the sql compiler prints it.